lsp.go

```diff
@@ -230,59 +225,72 @@ func (s *Server) handleDocumentSaved(
 	if err != nil {
 		log.Error().Err(err).Msg("reading saved file.")
 		return nil
 	}
 
-	err = checkFile(fname, string(data))
-	return reply(ctx, nil,
-		s.sendErrorDiagnostics(ctx, params.TextDocument.URI, err),
-	)
+	return s.checkAndReply(ctx, reply, fname, string(content))
 }
 
-func (s *Server) sendErrorDiagnostics(ctx context.Context, currentFile lsp.URI, err error) error {
-	if err == nil {
-		log.Debug().Str("file", currentFile.Filename()).
-			Msg("cleaning editor errors")
-
-		// this is required to clear the `problems panel` for the active file
-		// if it had errors before.
-		s.sendDiagnostics(ctx, currentFile, []lsp.Diagnostic{})
-		return nil
+func (s *Server) sendErrorDiagnostics(ctx context.Context, files []string, err error) error {
+	errs := errors.L()
+	switch e := err.(type) {
+	case *errors.Error:
+		// an error can wrap a list of errors.
+		errs = e.AsList()
+	case *errors.List:
+		errs = e
+	default:
+		if err != nil {
+			log.Debug().Err(err).Msg("unknown error ignored because it doesn't provide file range")
+		}
 	}
 
-	e, ok := err.(*errors.Error)
-	if !ok {
-		log.Debug().Err(err).Msg("unknown error ignored because it doesn't provide file range")
-		return nil
-	}
+	diagsMap := map[string][]lsp.Diagnostic{}
 
-	log.Debug().Str("error", e.Detailed()).Msg("sending diagnostics")
+	for _, err := range errs.Errors() {
+		e, ok := err.(*errors.Error)
+		if !ok {
+			log.Debug().Err(err).Msg("ignoring error without metadata")
+			continue
+		}
+
+		log.Debug().Str("error", e.Detailed()).Msg("sending diagnostics")
 
-	fileRange := lsp.Range{}
-	fileRange.Start.Line = uint32(e.FileRange.Start.Line) - 1
-	fileRange.Start.Character = uint32(e.FileRange.Start.Column) - 1
-	fileRange.End.Line = uint32(e.FileRange.End.Line) - 1
-	fileRange.End.Character = uint32(e.FileRange.End.Column) - 1
+		filename := e.FileRange.Filename
+		_, ok = diagsMap[filename]
+		if !ok {
+			diagsMap[filename] = []lsp.Diagnostic{}
+		}
+
+		fileRange := lsp.Range{}
+		fileRange.Start.Line = uint32(e.FileRange.Start.Line) - 1
+		fileRange.Start.Character = uint32(e.FileRange.Start.Column) - 1
+		fileRange.End.Line = uint32(e.FileRange.End.Line) - 1
+		fileRange.End.Character = uint32(e.FileRange.End.Column) - 1
 
-	// TODO(i4k): improve terramate to support multiple errors.
-	diags := []lsp.Diagnostic{
-		{
-			Message:  e.Error(),
+		diagsMap[filename] = append(diagsMap[filename], lsp.Diagnostic{
+			Message:  e.Message(),
 			Range:    fileRange,
 			Severity: lsp.DiagnosticSeverityError,
 			Source:   "terramate",
-		},
+		})
 	}
 
-	filePath := lsp.URI(uri.File(filepath.ToSlash(e.FileRange.Filename)))
-	s.sendDiagnostics(ctx, filePath, diags)
+	for _, filename := range files {
+		if _, ok := diagsMap[filename]; !ok {
+			diagsMap[filename] = []lsp.Diagnostic{}
+		}
+	}
 
-	if filePath != currentFile {
-		s.sendDiagnostics(ctx, currentFile, []lsp.Diagnostic{})
+	for _, filename := range files {
+		diags := diagsMap[filename]
```

Comment:
yeah it was explained in the other comment. We need to explicitly send a JSON containing:

```
{
    "diagnostics": []
}
```
and not:
```
{
    "diagnostics": null
}
```

The former is interpreted as **the file is ok, let's clear all the diagnostics in the problems panel**.
The latter is interpreted as **there's no diagnostic data in this message, just ignore**.

Reply:
https://github.com/mineiros-io/terramate-lsp/pull/6/files#diff-0a147895f2380ebd0028c790b6c2c47ec695e7f2387ab91d221b2685a4b678efR101

Reply:
we can handle the `!ok` in the beginning and send empty slices to all files, but all the rest of the code needs to still be there, because we need to send empty slices to all other files anyway.

Example:

- stack has 3 files: config.tm, globals.tm and stack.tm
- config.tm has errors but the other two are clean

so we need to send 3 events anyway:

1. config.tm list of errors
2. globals.tm empty slice
3. stack.tm empty slice

So even if we deal with `err == nil` or the `!ok` (if it's not a errors.E and errors.List), the rest of the code still need to be there ...

Reply:
so the current code handle all cases: `err is nil`, `err is not errors.E etc`, `err report errors for all files`, and `err report errors for only some files`.

Reply:
ah ok, I thought you're still talking about handling the empty slice in the beginning of the function. My bad..

Have a look in the last update in the code, now the empty slice is added to the map in just 1 place.